refactor, wallet: get serialized size of `CRecipient`s directly

josibake opened the pull request:
Broken out from #28201 

---

In order to estimate fees properly, we need to know what the final serialized transaction size will be. This PR refactors `CreateTransactionInternal` to:

* Get the serialized size directly from the `CRecipient`: this sets us up in a future PR to calculate the serialized size of silent payment `CTxDestinations` (see https://github.com/bitcoin/bitcoin/pull/28201/commits/797e21c8c1bf393e668eeef8bb7ee9cae1e5e41e)
* Use the new `GetSerializeSizeForRecipient` to move the serialize size calculation to *before* coin selection and the output creation to *after* coin selection: this also sets us up for silent payments sending in a future PR in that silent payments outputs cannot be created until after the inputs to the transaction have been selected

Aside from the silent payments use case, I think this structure logically makes more sense. As a reminder, move-only commits are best reviewed with something like `git diff -w --color-moved=dimmed-zebra`

josibake committed:
wallet: use CRecipient instead of CTxOut

Now that a CRecipient holds a CTxDestination, we can get the serialized
size and determine if the output is dust using the CRecipient directly.
This does not change any current behavior, but provides a nice generalization
that can be used to apply special logic to a CTxDestination serialization
and dust calculations in the future.

Specifically, in a later PR when support for `V0SilentPayment` destinations is
added, we need to use `WitnessV1Taproot` as the scriptPubKey for serialized
size calcuations whenever the `CRecipient` destination is a `V0SilentPayment`
destination.

josibake committed:
move-only: refactor CreateTransactionInternal

Move the output serialization size and dust calculation into the loop where the
outputs are iterated over to calculate the total sum.

Move the code for adding the the txoutputs to the transaction to after
coin selection.

While this code structure generally follows a more logical flow,
the primary motivation for moving the code for adding outputs to the
transaction sets us up nicely for silent payments (in a future PR):
we need to know the input set before generating the final output scriptPubKeys.

paplorinc commented on src/wallet/spend.cpp:
```diff
@@ -1141,6 +1129,16 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
            result.GetSelectedValue());
 
+    // vouts to the payees
+    for (const auto& recipient : vecSend)
```
would it make sense to preallocate the vouts before pushing?
```C++
    txNew.vout.reserve(txNew.vout.size() + vecSend.size());
```

josibake replied:
Trying to keep this a move-only refactor, so would prefer to leave it for now. I end up touching this code again in #28201; will make a note of this for that PR.

paplorinc replied:
Split it out to a new PR so that we don't forget it: https://github.com/bitcoin/bitcoin/pull/30093/files

paplorinc replied:
Rebased https://github.com/bitcoin/bitcoin/pull/30093 after this merge - it became trivial after your changes

paplorinc commented on src/wallet/spend.cpp:
```diff
@@ -994,6 +994,8 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
     // Set the long term feerate estimate to the wallet's consolidate feerate
     coin_selection_params.m_long_term_feerate = wallet.m_consolidate_feerate;
+    // Static vsize overhead + outputs vsize. 4 nVersion, 4 nLocktime, 1 input count, 1 witness overhead (dummy, flag, stack size)
+    coin_selection_params.tx_noinputs_size = 10 + GetSizeOfCompactSize(vecSend.size()); // bytes for output count
```
nit: found it a but confusing that we have two different style comments for the same line

josibake replied:
Would prefer to not change for now to keep this a move-only commit.

S3RK commented on src/wallet/spend.cpp:
```diff
@@ -1141,6 +1129,16 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
+    {
+        CTxOut txout(recipient.nAmount, GetScriptForDestination(recipient.dest));
+
+        if (IsDust(txout, wallet.chain().relayDustFee())) {
```
nit: it makes sense to keep this check before coin selection to return early and avoid wasting time

josibake replied:
Yeah, I had considered this but it would require refactoring `IsDust` in a non-trivial way, since `IsDust` is used outside of the wallet and as such doesn't really know what a `CRecipient` is.

It's definitely possible, since all you really need to check for dust is the serialized size of the output and the value, but doing the refactor here blows up the scope of this PR quite a bit. I think it would be better to handle this in a separate PR, i.e. refactor `IsDust` so it can be used for both `CRecipients` and `CTxOuts`. Happy to open a follow-up PR, or at the very least leave a `TODO` comment here.

josibake replied:
Actually, I just realized I was over complicating this: you don't need to refactor `IsDust`, you just need to wrap it with an `IsDust` function that turns a `CRecipient` into a `CTxOut`, same as `GetSerializeSizeForRecipient`. Will update.

josibake replied:
Update, thanks @S3RK !

josibake commented:
Updated https://github.com/bitcoin/bitcoin/commit/b44f00251a77a3360888eeb10ec903375eb1e0f8 -> https://github.com/bitcoin/bitcoin/commit/f004b1625bcb9b833f36af47b53675e3dd77135b ([serialize-size-for-crecipient-00](https://github.com/josibake/bitcoin/tree/serialize-size-for-crecipient-00) -> [serialize-size-for-crecipient-01](https://github.com/josibake/bitcoin/serialize-size-for-crecipient-01), [compare](https://github.com/josibake/bitcoin/compare/serialize-size-for-crecipient-00..serialize-size-for-crecipient-01))

* Added method for checking if `CRecipient` is dust directly
* This allows for checking the outputs are dust before running coin selection (h/t @S3RK )

S3RK commented:
Code Review ACK f004b1625bcb9b833f36af47b53675e3dd77135b

The PR doesn't modify the existing behavior, but prepares the code for Silent Payments (see #28201)
The code pushed down in the function doesn't affect the logic, because `txnew.vout` is not used between the old and the new location.

S3RK commented on src/wallet/wallet.cpp:
```diff
@@ -4559,4 +4559,18 @@ std::optional<CKey> CWallet::GetKey(const CKeyID& keyid) const
     return std::nullopt;
 }
+
+size_t GetSerializeSizeForRecipient(const CRecipient& recipient)
```
p.s. I don't see how the callable will be applied to more than one destination, so in my mind it's just easier to inline the callable and drop the `std::visit`

josibake replied:
Regarding the visitor pattern, `CTxDestination` is a std::variant and we need the visitor to know to use `WitnessV1Taproot` when calculating the serialized size for both `GetSerializeSizeForRecipient` and `IsDust` when the `CTxDestination` is a `V0SilentPayments` destination. See https://github.com/bitcoin/bitcoin/pull/28201/commits/797e21c8c1bf393e668eeef8bb7ee9cae1e5e41e.

However, you're correct this isn't really clear in this PR, so I'll update the commit message to explain this.

Regarding making the functions callable, I also wasn't sure about this. I noticed a few places where `IsDust` is used in the RPC calls, which is why I decided to make it callable here, but we can always remove them from the header if we end up only using it in `CreateTransactionInternal`.

josibake replied:
@S3RK , @furszy you're right: `get_if` works here, as well. For some reason I was thinking `std::get_if` doesn't check if the type exists in the `std::variant` at compile time, hence using a visitor was better. But `std::get_if` does check that the type is included in the variant; its just that the visitor can actually do compile time checks to ensure all the cases are covered. But in the silent payments example, we don't need exhaustive coverage so the visitor is likely overkill. I've updated the functions to remove the visitor.

I also moved these into `spend.cpp` for now. After looking around a bit, I don't think I will need them anywhere else and if they are, I can move them to a `spend_util.h/cpp` file, per @furszy 's suggestion.

josibake commented:
Rebased to fix merge conflict and update commit message to better explain the motivation for using the visitor patter, per @S3RK 's feedback

josibake commented on src/wallet/wallet.h:
```diff
@@ -293,6 +294,9 @@ struct CRecipient
     bool fSubtractFeeFromAmount;
 };
 
+size_t GetSerializeSizeForRecipient(const CRecipient& recipient);
+bool IsDust(const CRecipient& recipient, const CFeeRate& dustRelayFee);
+
```
IIRC, I tried to move `CRecipient` out of `wallet.h` awhile back and it ended up being more complicated than I expected? But in general I agree.

josibake commented on src/wallet/spend.cpp:
```diff
@@ -9,7 +9,6 @@
 #include <consensus/validation.h>
 #include <interfaces/chain.h>
 #include <numeric>
-#include <policy/policy.h>
```
Fixed.

josibake commented:
Updated to remove the visitor per @S3RK and @furszy 's feedback

S3RK commented:
reACK a9c7300135f0188daa5bce5491e2daf2dd8da8ae

Looks even simpler now, thanks for incorporating feedback.

rkrux commented on src/wallet/spend.cpp:
```diff
@@ -1098,23 +1106,6 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
-        if (IsDust(recipient, wallet.chain().relayDustFee())) {
-            return util::Error{_("Transaction amount too small")};
-        }
-        txNew.vout.push_back(txout);
```
Although being a bit pedantic, but as this commit is supposed to be move-only, the removal of `CTxOut txout...` in line 1107 and the replacement of `push_back` with `emplace_back` in line 1115 could have been part of the previous commit to avoid any confusion for future readers.

josibake replied:
Will leave for now so as to not require reACKs for a house keeping change, but will reorder if I end up needing retouch. Thanks!

rkrux reviewed (approved):
tACK [a9c7300](https://github.com/bitcoin/bitcoin/pull/30050/commits/a9c7300135f0188daa5bce5491e2daf2dd8da8ae)

`make, test/functional` are successful.

From what I understand about Silent Payments, the calculation of the recipient address is dependent on the input hash for which it makes sense to push down the `txoutputs` addition _after_ coin selection.

Left a suggestion regarding commits but not a blocker.

furszy reviewed (commented):
ACK functionality wise. Not blocking if others are happy with the current code.

Have to admit that I'm not fan of the multiple `CTxOut` creations and `GetScriptForDestination` calls when we could cache the `CTxOut` early in the process, leaving the silent payment destinations empty, just so they can be modified later on. The first commit `GetSerializeSizeForRecipient` and `IsDust` decouplings seem more like an unneeded indirection rather than something worth doing to me atm. But as said above, not blocking if others are happy with it.

achow101 commented:
ACK a9c7300135f0188daa5bce5491e2daf2dd8da8ae

> Have to admit that I'm not fan of the multiple `CTxOut` creations and `GetScriptForDestination` calls when we could cache the `CTxOut` early in the process, leaving the silent payment destinations empty, just so they can be modified later on

I don't like that since it overloads the meaning of an empty scriptPubKey. This would also not work as well for future silent payments versions since information about the version.

Shuhaib07 reviewed (commented):
Approve
